src/utils/capmodpp.cpp

```diff
@@ -0,0 +1,287 @@
+
+#include <platform.hpp>
+#include <iomanip>
+#include "capmodpp.hpp"
+
+
+#ifdef ANTINET_linux
+	#include <linux/types.h>
+
+	// uses libcap-ng
+	#include <cap-ng.h>
+#else
+	#error "not supported"
+#endif
+
+#define debug_capmodpp 0
+
+namespace capmodpp {
+
+const t_eff_value v_eff_enable{cap_permchange::enable};
+const t_eff_value v_eff_disable{cap_permchange::disable};
+const t_eff_value v_eff_unchanged{cap_permchange::unchanged};
+
+const t_permit_value v_permit_enable{cap_permchange::enable};
+const t_permit_value v_permit_disable{cap_permchange::disable};
+const t_permit_value v_permit_unchanged{cap_permchange::unchanged};
+
+const t_inherit_value v_inherit_enable{cap_permchange::enable};
+const t_inherit_value v_inherit_disable{cap_permchange::disable};
+const t_inherit_value v_inherit_unchanged{cap_permchange::unchanged};
+
+cap_nr get_last_cap_nr() {
+	return CAP_LAST_CAP;
+}
+
+cap_nr get_cap_size() {
+	return CAP_LAST_CAP+1;
+}
+
+std::string cap_nr_to_name(cap_nr nr) {
+	if (nr > CAP_LAST_CAP) {
+		std::ostringstream oss; oss << "invalid_cap_" << nr;
+		throw std::runtime_error(oss.str());
+	}
+	return capng_capability_to_name(nr);
+}
+
+cap_nr cap_name_to_nr(const std::string & name) {
+	auto nr = capng_name_to_capability(name.c_str());
+	long int nr_long{ nr };
+	long int last_long{ get_last_cap_nr() };
+	if ((nr_long >= last_long ) || (nr<0)) {
+		std::ostringstream oss; oss << "Can not convert cap name (" << name << ") to a valid cap number, got: " << nr;
+		throw std::runtime_error(oss.str());
+	}
+	return nr;
+}
+
+cap_state::cap_state() : eff(cap_perm::no), permit(cap_perm::no), inherit(cap_perm::no), bounding(cap_perm::no)
+{ }
+
+bool cap_state::is_usable() const {
+	if (this->eff == cap_perm::yes) return true;
+	if (this->permit == cap_perm::yes) return true;
+	if (this->inherit == cap_perm::yes) return true;
+	return false;
+}
+
+void apply_change_perm(cap_perm & perm , const cap_permchange & change) {
+	switch (change) {
+		case cap_permchange::enable: perm = cap_perm::yes; break;
+		case cap_permchange::disable: perm = cap_perm::no; break;
+		case cap_permchange::unchanged: break;
```

Comment:
:no_entry: (discussed) no need for default - instead, we rely on LACK of it to make compiler report warning when a possibke case is not handled.

eg
clang version 3.9.1-svn288847-1~exp1 (branches/release_39)

src/utils/capmodpp.cpp:244:10: warning: enumeration value 'unchanged' not handled in switch [-Wswitch]